apps/site/pages/en/learn/asynchronous-work/discover-promises-in-nodejs.md

```diff
@@ -0,0 +1,317 @@
+
+For example, the `fs` (file system) module has a Promise-based API under `fs.promises`:
+
+```js
+const fs = require('node:fs').promises;
+// Or, you can import the promisified version directly:
+// const fs = require('node:fs/promises');
+
+async function readFile() {
+  try {
+    const data = await fs.readFile('example.txt', 'utf8');
+    console.log(data);
+  } catch (err) {
+    console.error('Error reading file:', err);
+  }
+}
+
+readFile();
+```
+
+In this example, `fs.promises.readFile()` returns a Promise, which we handle using `async/await` syntax to read the contents of a file asynchronously.
+
+## Advanced Promise Methods
+
+JavaScript's `Promise` global provides several powerful methods that help manage multiple asynchronous tasks more effectively:
+
+### **`Promise.all()`**:
+
+This method takes an array of Promises and resolves them all. It only resolves when all Promises are fulfilled. If any of the Promises is rejected, `Promise.all()` will reject immediately.
+
+`Promise.all()` resolves when all Promises are completed. If you have a large number of Promises, especially in situations like batch processing, it could overwhelm the system's memory.
+
+```js
+const { setTimeout: delay } = require('node:timers/promises');
+
+const fetchData1 = delay(1000).then(() => 'Data from API 1');
+const fetchData2 = delay(2000).then(() => 'Data from API 2');
+
+Promise.all([fetchData1, fetchData2])
+  .then(results => {
+    console.log(results); // ['Data from API 1', 'Data from API 2']
+  })
+  .catch(error => {
+    console.error('Error:', error);
+  });
+```
+
+### **`Promise.allSettled()`**:
+
+This method waits for all promises to either resolve or reject and returns an array of objects that describe the outcome of each Promise.
+
+```js
+const promise1 = Promise.resolve('Success');
+const promise2 = Promise.reject('Failed');
+
+Promise.allSettled([promise1, promise2]).then(results => {
+  console.log(results);
+  // [ { status: 'fulfilled', value: 'Success' }, { status: 'rejected', reason: 'Failed' } ]
+});
+```
+
+Unlike `Promise.all()`, `Promise.allSettled()` does not short-circuit on failure. It waits for all promises to settle, even if some reject. This provides better error handling for batch operations, where you may want to know the status of all tasks, regardless of failure.
+
+### **`Promise.race()`**:
+
+This method resolves or rejects as soon as the first Promise settles, whether it resolves or rejects.
+
+```js
+const { setTimeout: delay } = require('node:timers/promises');
+
+const task1 = delay(2000).then(() => 'Task 1 done');
+const task2 = delay(1000).then(() => 'Task 2 done');
+
+Promise.race([task1, task2]).then(result => {
+  console.log(result); // 'Task 2 done' (since task2 finishes first)
+});
+```
```

Comment:
Because so many people get this wrong, it's worth pointing out that while the `Promise.race(...)` is resolved or rejected when the first promise resolves/rejects, the other promises are still going... that is, the other tasks are not canceled when one promise wins the race. I've lost count of the number of bugs I've run into in the wild because people didn't realize this.